crates/bevy_ecs/src/storage/table.rs

```diff
@@ -32,6 +32,37 @@ impl TableId {
     }
 }
 
+/// A opaque newtype for rows in [`Table`]s. Specifies a single row in a specific table.
+///
+/// Values of this type are retreivable from [`Archetype::entity_table_row`] and can be
+/// used alongside [`Archetype::table_id`] to fetch the exact table and row where an
+/// [`Entity`]'s
+///
+/// Values of this type are only valid so long as entities have not moved around.
+/// Adding and removing components from an entity, or despawning it will invalidate
+/// potentially any table row in the table the entity was previously stored in. Users
+/// should *always* fetch the approripate row from the entity's [`Archetype`] before
+/// fetching the entity's components.
+///
+/// [`Archetype`]: crate::archetype::Archetype
+/// [`Archetype::entity_table_row`]: crate::archetype::Archetype::entity_table_row
+/// [`Archetype::table_id`]: crate::archetype::Archetype::table_id
+/// [`Entity`]: crate::entity::Entity
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct TableRow(usize);
+
+impl TableRow {
+    #[inline]
+    pub(crate) const fn new(index: usize) -> Self {
+        Self(index)
+    }
+
+    #[inline]
+    pub(crate) fn index(self) -> usize {
```

Comment:
```suggestion
    pub fn index(self) -> usize {
```
I also don't really see a problem with letting users know what the actual index is

Reply:
Done.